### PR TITLE
[embedlite-components] Provide input context info for predictive text input. Contributes to JB#55264

### PR DESCRIPTION
**Pull request by paveltkv**

### CODE REVIEW

**Comment by paveltkv** [comment]
The full set of PRs related to this:
https://github.com/sailfishos/embedlite-components/pull/42 (this)
https://github.com/sailfishos/gecko-dev/pull/56
https://github.com/sailfishos/qtmozembed/pull/9

File: jsscripts/Makefile.am [hunk 1]
```diff
@@ -4,6 +4,7 @@ jsscripts_manifest_DATA = \
 	Logger.js \
 	Feeds.jsm \
 	FormAssistant.js \
+        InputMethodHandler.js \
```
Comment (okodron) [inline jsscripts/Makefile.am:7]: Something is wrong with formatting/padding.

**Review by okodron — APPROVED** [review]

File: jsscripts/InputMethodHandler.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    try {
+      let winId = Services.embedlite.getIDByWindow(aElement.ownerGlobal);
+      Services.embedlite.sendAsyncMessage(winId, "InputMethodHandler:ResetInputContext", "[]");
+    } catch (e) {
+      Logger.warn("InputMethodHandler: sending async message failed", e);
+    }
+  },
+
+  // We only want to show login suggestions for certain elements
+  _isAutoComplete: function(aElement) {
```
Comment (rainemak) [inline jsscripts/InputMethodHandler.js:115]: Smells like a copy-paste from FormAssistant.js. I'd recommend different kind of composition.

There are two options I think

- Add the beef of this one to the FormAssistant.js. Question then being that we should that FormAssistant.js be called?
- Alternatively we could think of splitting FormAssistant into two and use common part from both FormAssistant.js and InputMethodHandler.js.

I do not have a strong take yet which is better way as both could potentially be good.

**Review by rainemak — CHANGES_REQUESTED** [review]
Added a review comment already regarding the copy-paste code.

File: jsscripts/InputMethodHandler.js (hunk 2 above)
Comment (paveltkv) [reply jsscripts/InputMethodHandler.js:115]: Yep, this is copy-paste, but the implementation of _isAutoComplete is slightly different from the one in FormAssistant.js. About comment line, I missed it. It should be changed or removed. The second is preferable in this context I think.

Comment (rainemak) [reply jsscripts/InputMethodHandler.js:114]: Alright. Created a follow up task to sync/align copy-pasted parts (JB#55725) - I think we can live with this for time being.

File: jsscripts/InputMethodHandler.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Copyright (c) 2021 Open Mobile Platform LLC.
+ */
+
+"use strict";
+
+var globalObject = null;
```
Comment (rainemak) [inline jsscripts/InputMethodHandler.js:10]: I'd call this one gInputMethodHandler instead.

**Review by rainemak — APPROVED** [review]

**Pull request merged**